playground/strategic_bot.py

```diff
@@ -0,0 +1,223 @@
+from util import turn_check, Direction
+
+
+@dataclass
+class Edge:
+	stone: Stone
+	directions: List[str]
+
+
+class StrategicBot:
+	def __init__(self, board_config: BoardConfig):
+		self.k = board_config.connect
+		self.m = board_config.row
+		self.n = board_config.column
+		self.p = board_config.each_move
+		self.q = board_config.first_move
+
+		self.directions = {
+			Direction.RIGHT: lambda x, y: (x + 1, y),
+			Direction.DOWN: lambda x, y: (x, y + 1),
+			Direction.DOWN_RIGHT: lambda x, y: (x + 1, y + 1),
+			Direction.UP_RIGHT: lambda x, y: (x + 1, y - 1)
+		}
+		self.counter_directions = {
+			Direction.RIGHT: lambda x, y: (x - 1, y),
+			Direction.DOWN: lambda x, y: (x, y - 1),
+			Direction.DOWN_RIGHT: lambda x, y: (x - 1, y - 1),
+			Direction.UP_RIGHT: lambda x, y: (x - 1, y + 1)
+		}
+
+		self.opponent_color = ""
+		self.my_color = ""
+
+	# edge is left most or upper most stone in a connection size greater than 1: left has priority
+	def is_stone_edge(self, stone: Stone, log: List[Stone]) -> (bool, List[str]):
+		default_directions = [Direction.RIGHT, Direction.DOWN, Direction.DOWN_RIGHT, Direction.UP_RIGHT]
+		directions = []
+		for dd in default_directions:
+			counter_f = self.counter_directions[dd]
+			m_x, m_y = counter_f(int(stone.x), int(stone.y))
+			m_stone = Stone(str(m_x), str(m_y), stone.color)
+			if m_stone in log:
+				break
+			f = self.directions[dd]
+			m_x, m_y = f(int(stone.x), int(stone.y))
+			m_stone = Stone(str(m_x), str(m_y), stone.color)
+			if m_stone in log:
+				directions.append(dd)
+		if len(directions) > 0:
+			is_edge = True
+		else:
+			is_edge = False
+		return is_edge, directions
+
+	# TODO: to be generalized to util (on the next phase)
+	def connection_check(self, edge: Edge, log: List[Stone], num: int):
+		is_connected = False
+		filtered_direction = {}
+		for direction_name, direction in self.directions.items():
+			if direction_name in edge.directions:
+				filtered_direction[direction_name] = direction
+		for _, direction in filtered_direction.items():
+			connection_count = 1
+			new_x, new_y = (edge.stone.x, edge.stone.y)
+			while True:
+				new_x, new_y = direction(int(new_x), int(new_y))
+				new_stone = Stone(str(new_x), str(new_y), edge.stone.color)
+				if new_stone in log:
+					connection_count += 1
+				else:
+					break
+			if connection_count == num:
+				is_connected = True
+				break
+		return is_connected
+
+	def group_by_connected(self, log: List[Stone], color: str) -> Dict[int, List[Edge]]:
+		groups = defaultdict(list)
+		edges = defaultdict(list)
+		for s in log:
+			is_edge, directions = self.is_stone_edge(s, log)
+			if s.color == color and is_edge:
+				edges[(s.x, s.y, s.color)] = directions
+		# returns connected that are not closed
+		for edge_, dirs in edges.items():
+			edge = Edge(Stone(edge_[0], edge_[1], edge_[2]), dirs)
+			for num in list(range(6, 1, -1)):
+				if self.connection_check(edge, log, num):
+					groups[num].append(edge)
+					break
```

Comment:
could early return?

Reply:
can't return in current loop b/c it has double for loops

Reply:
got it, thanks